src/plugins/discover/public/context_awareness/profiles/data_source_profile.ts

```diff
@@ -0,0 +1,45 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+import type { DataView } from '@kbn/data-views-plugin/common';
+import type { AggregateQuery, Query } from '@kbn/es-query';
+import type { DiscoverDataSource } from '../../../common/data_sources';
+import { AsyncProfileService } from '../profile_service';
+import { Profile } from '../types';
+
+export enum DataSourceCategory {
+  Logs = 'logs',
+  Default = 'default',
+}
+
+export interface DataSourceProfileProviderParams {
```

Comment:
> [!TIP]  
> Can we make this a discriminated union?
```ts
export type DataSourceProfileProviderParams =
  | {
      dataSource: DataViewDataSource;
      dataView?: DataView;
    }
  | {
      dataSource: EsqlDataSource;
      query?: Query | AggregateQuery;
    };
```

Reply:
I like the idea and I explored it for a bit, but it doesn't seem like type narrowing works as one might expect here. It seems to be a similar problem as the one mentioned here: https://stackoverflow.com/questions/70347606/why-doesnt-typeof-narrow-a-union-type-when-the-type-of-a-property-is-a-discri.

I'm open to suggestions though if you see something I'm missing or can get it working.